New config: handles both old and new config files

bbaldino opened the pull request:
This is still very much a WIP.  After thinking about ways to convert old config files to new ones, I thought we could model this logic in classes for the properties themselves.  This has some other advantages like giving us a good solution for future changes to config (naming changes, semantic changes, etc) without breaking things.  Using this, I think we can show warnings about deprecated/no-longer-used config values as well.

bbaldino commented:
Some more info on the concept:

Every configuration property has a corresponding class.  This class can encapsulate logic such as its old legacy name and its new name.  Getting the value uses something called a 'retriever': a retriever is given a config to read from and a property key to check for.  In our case to start, most configuration properties will have 2 retrievers: the first will check the old config using the old property name and the second will check the new config with the new property name.  Retrievers can do more as well, for example if the semantics of a property are different between the 2 config files (say the old one was `XXX_DISABLED` and the new one is `XXX_ENABLED`), the retriever can also perform a transformation on the retrieved value (in this case, we'd invert the value retrieved from the old config to match the new semantics). 

There's a lot of code to power all this (though it is definitely a WIP, I think there's some cleanup that can be done), the actual usage isn't too bad (and can likely be improved).  Here's an example property definition:

```
public class HealthTimeoutProperty
{
    protected static final String legacyPropKey = "org.jitsi.videobridge.health.TIMEOUT";
    protected static final String propKey = "videobridge.health.timeout";

    static ConfigProperty<Integer> createInstance()
    {
        return new ConfigPropertyBuilder<Integer>()
                .usingGetter((config, key) -> (int)config.getDuration(key, TimeUnit.MILLISECONDS))
                // We'll use the value from legacy config if it's present, then try the new config
                .fromConfigs(
                        new DefaultLegacyConfigValueRetrieverBuilder<>(legacyPropKey)
                        new DefaultConfigValueRetrieverBuilder<>(propKey),
                )
                // This prop only gets read once, i.e. it won't update based on config file changes
                .readOnce()
                .build();
    }

    private static ConfigProperty<Integer> singleInstance = createInstance();

    public static ConfigProperty<Integer> getInstance()
    {
        return singleInstance;
    }
}
```

bbaldino commented on src/main/java/org/jitsi/videobridge/health/Health.java:
```diff
@@ -286,26 +251,7 @@ private static String generateEndpointID()
-        int period =
-            cfg == null ? PERIOD_DEFAULT
-                : cfg.getInt(PERIOD_PNAME, PERIOD_DEFAULT);
-        setPeriod(period);
```
this is no longer needed.  before, on line 289, we invoked super with a default period and then we'd get the property value and override the one we'd previously set.  now we just get the value directly in the super call (which will have the fallback-to-default built in)

bbaldino replied:
i do think that was subtle before, i was confused by it when i saw it.  hopefully it's clearer now.

JonathanLennox replied:
Ah, I see, I missed it in the super constructor. I agree this is better.

bbaldino commented:
Notes on some changes since my last comment:

1) I moved the generic bits out to jitsi-utils.  That code can now be found in a draft PR [here](https://github.com/jitsi/jitsi-utils/pull/23)
2) Retrievers are now gone.  I realized we could use `Supplier` and get basically the same thing so I changed things to just use `Supplier`.  This also an advantage in that it was easier to move some useful, generic parts (which have no knowledge of the config library) out to jitsi-utils.
3) I added some code to start playing with logging warnings for the use of 'obsolete' configuration properties (see [DummyObsoleteProperty](https://github.com/jitsi/jitsi-videobridge/blob/d7d7ed2592193c805fa6cede378c49cde5844b47/src/main/java/org/jitsi/videobridge/rest/DummyObsoleteProperty.java) and [validateConfig--still a WIP](https://github.com/jitsi/jitsi-videobridge/blob/d7d7ed2592193c805fa6cede378c49cde5844b47/src/main/java/org/jitsi/videobridge/Main.java#L175).  The idea here is: when we make a property obsolete, we'd leave its property class in the code marked with an `ObsoleteConfig` annotation for some amount of time.  Main now has some code that will find all `ConfigProperty` instances marked with `ObsoleteConfig` and see if they have a value defined in configuration files; if they do, we  log a warning.  I think this approach could be extended to include checking for properties defined in config that we're not reading anywhere (to catch typos or stale properties).

bbaldino commented:
superseded by https://github.com/jitsi/jitsi-videobridge/pull/935
